### PR TITLE
feat: add datastore onSubmitBefore and onSubmitComplete callbacks

### PR DESCRIPTION
**Pull request by scottyoung**
*Issue #, if available:* 
https://app.asana.com/0/1202624607923297/1202422922235645

*Description of changes:* 
listen to onSubmitBefore and onSubmitComplete events 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap [hunk 1]
```diff
@@ -256,7 +288,23 @@ export default function myPostForm(props) {
     <form
       onSubmit={async (event) => {
         event.preventDefault();
+        if (onSubmitBefore) {
+          onSubmitBefore({ fields: myPostFormFields });
+        }
         myPostFormOnSubmit();
```
Comment (Jshhhh) [inline packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap:294]: Should be removed

Comment (Jshhhh) [inline packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap:294]: Should be removed also

File: packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap [hunk 2]
```diff
@@ -125,7 +141,23 @@ export default function myPostForm(props) {
...
```
[the remark this answers is not in the document]
Comment (scottyoung) [reply packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap]: done in next pr

File: packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap [hunk 3]
```diff
@@ -21,7 +21,23 @@ export default function customDataForm(props) {
     <form
       onSubmit={async (event) => {
         event.preventDefault();
+        if (onSubmitBefore) {
+          onSubmitBefore({ fields: customDataFormFields });
```
[the remark this answers is not in the document]
Comment (SwaySway) [reply packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap:25]: Maybe we can conditionally update the state if `onSubmitBefore` is updated?

[the remark this answers is not in the document]
Comment (scottyoung) [reply packages/codegen-ui-react/lib/__tests__/__snapshots__/studio-ui-codegen-react-forms.test.ts.snap]: updated

File: packages/codegen-ui-react/lib/amplify-ui-renderers/form.ts [hunk 4]
```diff
@@ -49,6 +48,7 @@ export default class FormRenderer extends ReactComponentRenderer<BaseComponentPr
     );
 
     this.importCollection.addImport('@aws-amplify/ui-react', this.component.componentType);
```
Comment (Jshhhh) [inline packages/codegen-ui-react/lib/amplify-ui-renderers/form.ts:50]: useTypeCastFields needs to be added to @aws-amplify/ui-react import

Comment (scottyoung) [reply packages/codegen-ui-react/lib/amplify-ui-renderers/form.ts:50]: just pushed an update to include that in the import, thanks

**Review by SwaySway — APPROVED** [review]
lgtm

**Review by Jshhhh — APPROVED** [review]

**Pull request merged**